BLE Indoor Positioning/src/main/java/com/nexenio/bleindoorpositioning/ble/advertising/AdvertisingPacketUtil.java

```diff
@@ -55,4 +57,90 @@ public static float getPacketFrequency(int packetCount, long duration, TimeUnit
         return packetCount / (float) (timeUnit.toSeconds(duration));
     }
 
+    /**
+     * Returns an ArrayList of AdvertisingPackets that have been received in the specified time range.
+     * If no packets match, an empty list will be returned. Expects the specified AdvertisingPackets
+     * to be sorted by timestamp in ascending order (i.e. the oldest timestamp first)!
+     *
+     * @param advertisingPackets the packets to filter (sorted by timestamp ascending)
+     * @param startTimestamp     minimum timestamp, inclusive
+     * @param endTimestamp       maximum timestamp, exclusive
+     */
+    public static <P extends AdvertisingPacket> ArrayList<P> getAdvertisingPacketsBetween(final ArrayList<P> advertisingPackets, long startTimestamp, long endTimestamp) {
+        // check if advertising packets are available
+        if (advertisingPackets.isEmpty()) {
+            return new ArrayList<>();
+        }
+
+        P oldestAdvertisingPacket = advertisingPackets.get(0);
+        P latestAdvertisingPacket = advertisingPackets.get(advertisingPackets.size() - 1);
+
+        // check if the timestamps are out of range
+        if (endTimestamp <= oldestAdvertisingPacket.getTimestamp() || startTimestamp > latestAdvertisingPacket.getTimestamp()) {
+            return new ArrayList<>();
+        }
+
+        P midstAdvertisingPacket = advertisingPackets.get(advertisingPackets.size() / 2);
+
+        // find the index of the first advertising packet with a timestamp
+        // larger than or equal to the specified startTimestamp
+        int startIndex = 0;
+        if (startTimestamp > oldestAdvertisingPacket.getTimestamp()) {
+            // figure out if the start timestamp is before or after the midst advertising packet
+            ListIterator<P> listIterator;
+            if (startTimestamp < midstAdvertisingPacket.getTimestamp()) {
+                // start timestamp is in the first half of advertising packets
+                // start iterating from the beginning
+                listIterator = advertisingPackets.listIterator();
+                while (listIterator.hasNext()) {
+                    if (listIterator.next().getTimestamp() >= startTimestamp) {
+                        startIndex = listIterator.nextIndex() - 1;
+                        break;
+                    }
+                }
+            } else {
+                // start timestamp is in the second half of advertising packets
+                // start iterating from the end
+                listIterator = advertisingPackets.listIterator(advertisingPackets.size());
+                while (listIterator.hasPrevious()) {
+                    if (listIterator.previous().getTimestamp() < startTimestamp) {
+                        startIndex = listIterator.nextIndex() + 1;
```

Comment:
Should be `startIndex = listIterator.nextIndex()` or `startIndex = listIterator.previousIndex() + 2;`.

Reply:
This will break the tests, e.g.:

```
java.lang.AssertionError: Packet timestamp before minimum timestamp
	at org.junit.Assert.fail(Assert.java:88)
	at com.nexenio.bleindoorpositioning.ble.beacon.BeaconTest.getAdvertisingPacketsSince(BeaconTest.java:114)
```